### ISSUE TITLE
Google Oauth keeps asking user consent

### ISSUE THREAD

**Issue by yannrouillard** [rancher/rancher#23633]
**What kind of request is this (question/bug/enhancement/feature request):**

bug

**Steps to reproduce (least amount of steps as possible):**

Configure Google Oauth following the official documentation.

**Result:**

Each time a user tries to login using Google Oauth, he is asked to validate the application access with the permissions _Display domain groups_ and _Display domain users_ being displayed.

It might be related to the wide permissions being asked, I don't understand why this keep being asked, I though this was not necessary thanks to the service account being created.

**Other details that may be helpful:**

Another possibly related issue is that if one log using local account, he can't see properly Google logged user information, nor can't he add to a group a Google Account as a member.

It seems Rancher always use the token of the logged user to get the information instead of the service account, so he can't do anything related to Google user/group when someone doesn't log with Google Oauth.

**Environment information**
- Rancher version: 2.3.0
- Installation option: single install

**Comment by mrajashree** [comment]
@yannrouillard This is not a bug. We require a refresh token from google when the user logs in so that we can refresh their  access token. And the request done for this involves asking user read-only consent to users/groups.
And when you login as local user you cannot see any external auth provider related info, same applies to google oauth

**Comment by yannrouillard** [comment]
@mrajashree So this is not a bug, this is a feature :)

So maybe this bug can be turned into documentation update if I am not the only one to not find this very clear.

In https://rancher.com/docs/rancher/v2.x/en/admin-settings/authentication/google/#3-creating-service-account-credentials it was written:

> Since the Google Admin SDK is available only to admins, regular users cannot use it to retrieve profiles of other users or their groups. Regular users cannot even retrieve their own groups.
> [...]
> As a workaround to get this capability, G Suite recommends creating a service account and delegating authority of your G Suite domain to that service account.

Reading at this, I assumed that regular users can't get these information so a service account will be used to gather that information instead of using the user token, which is clearly not the case.

Can I ask how is the service account used then ? You use it in conjunction with the user token ?

Also a warning about the fact that the user consent will be asked each time would be nice.

**Comment by yannrouillard** [comment]
Also when I was on this page https://cloud.google.com/iam/docs/understanding-service-accounts:

> A service account is a special type of Google account intended to represent a non-human user that needs to authenticate and be authorized to access data in Google APIs.
> Your application assumes the identity of the service account to call Google APIs, so that the users aren't directly involved.

So I was really under the impression that service account can be used without requiring any user token at all.

**Comment by yannrouillard** [comment]
> And when you login as local user you cannot see any external auth provider related info, same applies to google oauth

That is really weird. Nothing prevents you for showing them in a LDAP configuration for instance.

**Comment by phillycheeze** [comment]
Also seeing the same issue. The consent screen should only be happening once, correct? Maybe I am misunderstanding the flow of how Rancher does SSO with Google, but I've seen multiple implementations of service accounts in regards to delegating admin permissions that would never require a consent screen each time for every user.

@yannrouillard this is not the expected behavior, so I don't think rancher should update documentation to warn users about it. This feature was added in v2.3 and I assume not very many people are using it yet. I've let my engineering team know that the popup will hopefully be fixed in a future version, but if not, we will probably just switch to local authentication.

**Comment by phillycheeze** [comment]
The reason we would switch to local authentication is because prompting this consent for each login is a bad security practice. Training users to blindly hit "Allow" on each login will enable a terrible habit (i.e. "Alert Fatigue").

That popup is also enabling permissions for the _service account_, not the _user_. Why are we having normal users "allowing" an admin account to access admin features? An already existing Google Admin creates the service account, which Rancher on the backend should use whenever it needs to fetch the domain users or groups. The service account is just a proxy it uses to get admin-specific features from a domain, which is why a Google Admin must create the service account for Rancher.

**Comment by mrajashree** [comment]
@phillycheeze Like I mentioned in a previous comment, along with an access token we also need a refresh token for the user. The access token may expire and refresh token is used to obtain a new one. We use the access_type=offline option to request this refresh token. But google oauth sends this request token only the first time your application exchanges an authorization code for a user. 
After this, on your google account's permissions page, rancher will be listed as an authorized app. Unless you remove it from that, the next time user has to login, the refresh token won't be sent.
Prompting user consent is a way around it to get a refresh token on each login. It may be inconvenient but I don't see how it is a bad security practice?
Please refer to this for more details: https://stackoverflow.com/questions/10827920/not-receiving-google-oauth-refresh-token

I

**Comment by phillycheeze** [comment]
@mrajashree Yes, this is for the OAuth consent flow. What we are talking about is different.

After getting the refresh token, Rancher prompts all users a second time to grant access to users and groups. This flow is separate from the OAuth. Rancher does not need this authorization again, once the initial administrator has set up the service account.

**Comment by phillycheeze** [comment]
![image](https://user-images.githubusercontent.com/6655673/68321692-c7bf2800-0087-11ea-8a40-46d7e844331a.png)

Please see the screenshot above. This consent screen pops up everytime, for every user. This should not be prompted. The refresh token for the user is done one screen before this, where the user is asked which account they would like to use to login or possibly re-entering their Google password.

**Comment by mrajashree** [comment]
This screen is the one from which we get refresh tokens on subsequent logins. Without this you get the refresh token on the very first login with your app (Rancher) but not on any subsequent logins. It is being shown in response to the prompt=consent query attribute we use.

**Comment by phillycheeze** [comment]
I guess maybe I'm confused on how Rancher does the flow. Why would Rancher need a normal user's refresh or access token to access their domain's users/groups? Rancher, when needing access to a domain's users/groups, should instead use the service account behind the scenes to get this information. Not prompt the user for these permissions. 

Non-admin users in Rancher should not be granting the Rancher server access to the google admin sdk via this prompt.

**Comment by phillycheeze** [comment]
The answers on stackoverflow are talking about getting a new refresh_token, which Rancher only needs one time. When an admin sets up Google SSO in the Rancher settings, you make an initial prompt to that admin user to authorize the service account. Rancher should then store the refresh_token in the database.

All subsequent API calls can use that refresh_token to get a new access_token if the old access_token expires. The users shouldn't need to be prompted more than once, and the service account authorization should especially not be shown more than once to the original administrator user. 

Maybe I need more context on why Rancher needs a new refresh_token every time. That kind of goes against the basic mechanisms of how OAuth works.

**Comment by yannrouillard** [comment]
@phillycheeze I totally agree with your views and honestly I talked about documentation update in case I was wrong on this, but what I intended to do was to look how to modify the implementation to have the proper workflow, if there were no progress on this bug.

**Comment by yannrouillard** [comment]
@mrajashree If think there 2 different issues:

- **the fact that Rancher requires consent each time**
I am currently looking at the link you sent me to understand your exact issue but what I know for sure is that it is the first application that I see with this behaviour. All other applications hooked to Google Oauth ask my consent only once, so I suspect there is something wrong in the Rancher implementation.

- **the fact that Rancher ask for domain group/users view permissions**
It is also weird since the Service Account is supposed to be used for that purpose, can you explain us how the Service Account is currently being used and how an users consenting to domain group/users view permissions will help the Service Account do its job?

**Comment by mrajashree** [comment]
@phillycheeze 
> When an admin sets up Google SSO in the Rancher settings, you make an initial prompt to that admin user to authorize the service account.
I know that. An alternative to storing each user's refresh token was prompting consent. Storing refresh token can be looked into however prompting consent is still not a bad security practice.

> Maybe I need more context on why Rancher needs a new refresh_token every time. That kind of goes against the basic mechanisms of how OAuth works.
I already gave a detailed explanation. It does not go agains the basic mechanism

@yannrouillard 
> It is also weird since the Service Account is supposed to be used for that purpose, can you explain us how the Service Account is currently being used and how an users consenting to domain group/users view permissions will help the Service Account do its job?
This is not an issue and not something that has alternatives.
There is detailed explanation about this in [docs](https://rancher.com/docs/rancher/v2.x/en/admin-settings/authentication/google/#3-creating-service-account-credentials)
You can refer google documentation too https://developers.google.com/admin-sdk/directory/v1/guides/delegation

**Comment by phillycheeze** [comment]
I guess we fundamentally disagree. I've set up this exact flow before with internal applications, and there is no reason to ask for consent each time. The flow I described above is how it should be handled. 

If this isn't seen as a bug, that's very unfortunate and concerning. I'll continue to monitor this issue and see if anyone else can potentially help with the problem. In the meantime, I'll switch back to local auth.

**Comment by vincent99** [comment]
@mrajashree Sorry but I have to agree the behavior implemented is not good/"wrong".  Users shouldn't be prompted again to give permission they've already given us;  We should be storing the refresh token.

**Comment by mrajashree** [comment]
@phillycheeze I already said the option of storing refresh tokens can be looked into and have a separate issue tracking it. https://github.com/rancher/rancher/issues/23954

**Comment by phillycheeze** [comment]
Awesome, thanks! I might've misunderstood before. Thanks for creating the new issue, really appreciate it!

**Comment by mrajashree** [comment]
@phillycheeze @vincent99 Sure, I was earlier not sure why prompting consent was bad or if could be avoided, but the storing refresh token approach would solve that. 
I created a separate issue to track this though because this issue contains many other questions including whether or not logging in as a local account should show principals from external auth providers, which isn't specific to google oauth. Or the usage of service account which I still see no alternative to.
So I'm closing this issue and tracking the storing refresh token option with this issue https://github.com/rancher/rancher/issues/23954

**Comment by phillycheeze** [comment]
Again, maybe I missed something above, but this is two distinct issues. Normal users should never see the domain prompt. The one I added an image to above. This is a one time prompt to the administrator user, that should never be prompted again. Hopefully that can be addressed in this same issue as well.

The second issue being refresh tokens and forced prompts which you linked the new issue to, thanks for creating that!

**Comment by mrajashree** [comment]
Oh I think the prompt=consent query attribute we use is the one showing this consent screen. When we switch to storing refresh token, we won't need to use that attribute and that prompt won't be presented to normal user from my understanding.

**Comment by yannrouillard** [comment]
@mrajashree Thanks for creating the other issue. I don't mind closing this issue since your ticket will solve the original issue stated in the title.

However I agree with @phillycheeze that it should not solve the fact that Rancher should not have to ask for Domain wide permissions to the user.
And looking at the link you sent us: https://developers.google.com/admin-sdk/directory/v1/guides/delegation, it explicitely says:
> In enterprise applications you may want to programmatically access a user's data *without any manual authorization on their part*.

I can create a separate issue for that if you want.

And I can also create a separate issue for the `logging in as a local account should show principals from external auth providers` since it seems to be a general issue not tied to Google Oauth.

**Comment by mrajashree** [comment]
@yannrouillard The consent screen won't be shown on next logins. 
> In enterprise applications you may want to programmatically access a user's data without any manual authorization on their part.

This is what the service account does, and apart from the logic to prompt consent screen (which will be removed), I don't there's any other explicit manual authorization right?

> And I can also create a separate issue for the logging in as a local account should show principals from external auth providers since it seems to be a general issue not tied to Google Oauth.

Sure, although since this is true for all of rancher's auth providers, there may already be a similar issue, I'm not able to find it right now

**Comment by yannrouillard** [comment]
> This is what the service account does, and apart from the logic to prompt consent screen (which will be removed), I don't there's any other explicit manual authorization right?

I think you would always have a consent screen at the first login in Rancher2, I think it is the case, it would be normal that I have to agree at least one time before giving my information (email and name at least) to Rancher application.
Anyway, I can wait to see if #23954 solve this issue before opening another ticket :)

> Sure, although since this is true for all of rancher's auth providers, there may already be a similar issue, I'm not able to find it right now

Didn't find it either so created one here: #23955

**Comment by yannrouillard** [comment]
@mrajashree Sorry just an update on the issue **Rancher ask for domain group/users view permissions**

I just successfully implemented a script that retrieves group information and memberships from Google Directory API without needing any kind of user consent at all.

So I am still puzzled about why you need to add the _domain group_ and _domain users_ permissions in the user consent screen.

Do I open another bug about that ?

**Comment by phantasia15** [comment]
@yannrouillard @mrajashree 
Hi, I'm digging a bit into the code and seems like the Rancher UI is appending group/users view permission to the url when redirecting to Google page.  
I've tried to remove those permisisons from [this](https://github.com/rancher/ui/blob/master/lib/shared/addon/oauth/service.js#L6) line and now the user consent screen doesn't requests unnecessary permissions anymore while the rancher server still correctly gets the group info using the ServiceAccount.  

So perhaps this might be a bug in the UI? Should I open a PR to fix this?  

As the way it is now, group/user view permissions are classified as sensitive scopes so the consent screen will show "unverified app" and there is a cap on how many users (100)) can be authenticated, unless you are going through the app verification process, which can take quite a long time.

**Comment by phillycheeze** [comment]
A PR would be awesome @phantasia15 . Our org turned off Google sign-in because these bugs make it super annoying to use.

**Comment by phantasia15** [comment]
> 
> 
> A PR would be awesome @phantasia15 . Our org turned off Google sign-in because these bugs make it super annoying to use.

I'm not completely sure if this is a bug or some intentional design that I'm not aware of so I'll wait for the confirmation from Rancher team before opening a PR.  
In the mean time, you can test it by building a custom docker image like this:   

```
FROM rancher/rancher
RUN sed -i 's/https:\/\/www.googleapis.com\/auth\/admin.directory.user.readonly https:\/\/www.googleapis.com\/auth\/admin.directory.group.readonly//g' /usr/share/rancher/ui/assets/*.js

```
Do remember to clear your browser cache so the new js files are loaded.
